### PR TITLE
Add recipe support

### PR DESCRIPTION
**Pull request by Pikachu920**
### Description
Adds recipe creation section like this:
```
on script load:
  create a crafting recipe:
    key: "dirty-diamond"
    result: diamond named "Dirty Diamond"
    ingredients:
      dirt, dirt and dirt
      dirt, diamond and dirt
      dirt, dirt and dirt
	  
  create a furnace recipe:
    key: "dirty-diamond-boots"
    result: diamond boots named "Dirty Diamond Boots"
    ingredient: diamond
    cook time: 1 second
    xp: 500
	
  create a campfire recipe:
    key: "chainmail-boots"
    result: chainmail boots
    ingredient: diamond boots
    cook time: 1 second
    xp: 500
```

---
**Target Minecraft Versions:** Any
**Requirements:** N/A
**Related Issues:** probably but I'll find them later

### CODE REVIEW

File: src/main/java/ch/njol/skript/sections/SecRecipe.java [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+				Skript.error("A recipe line may only contain a list of items");
+				return false;
+			}
+			if (((ExpressionList<ItemType>) ingredients).getExpressions().length != 3) {
+				Skript.error("An recipe line may only contain three items");
+			}
+			List<TriggerSection> currentSections = getParser().getCurrentSections();
+			if (currentSections.isEmpty())
+				return false;
+			return currentSections.get(currentSections.size() - 1).getClass() == SecRecipe.class;
```
Comment (Pikachu920) [inline src/main/java/ch/njol/skript/sections/SecRecipe.java:133]: check the section first. we don't want to print errors if we're in the wrong section

File: src/main/java/ch/njol/skript/sections/SecRecipe.java [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+		key = (Expression<String>) exprs[1];
+		loadCode(sectionNode);
+		if (first == null) {
+			Skript.error("A recipe section must contain 3 recipe lines");
+			return false;
+		}
+		TriggerItem current = first;
+		for (int i = 0; i < 3; i++) {
+			if (current == null || current.getClass() != EffRecipeLine.class) {
+				Skript.error("A recipe section may only contain 3 recipe lines - 1");
```
Comment (Pikachu920) [inline src/main/java/ch/njol/skript/sections/SecRecipe.java:60]: remove debug numbers from the end of these errors

**Comment by Ankoki** [comment]
Just a suggestion but do you think it would be worth this being a structure when they're released? I feel like as these aren't dependent on any event these can be utilised.

**Comment by Fusezion** [comment]
> Just a suggestion but do you think it would be worth this being a structure when they're released? I feel like as these aren't dependent on any event these can be utilised.

> https://discord.com/channels/135877399391764480/836220422223036467/1041449280776458300

I believe there was talk about whether they should be made as a structure but from what I remember we mostly put it down
not fully sure what the reasons were

**Comment by Ankoki** [comment]
> I believe there was talk about whether they should be made as a structure but from what I remember we mostly put it down not fully sure what the reasons were

I see the reasoning here tbh, however arent structures able to be used as sections? i could be missing the ballpark completely but supporting both seems like something that structure is in range for, and if not that would be a nice imrpovment to structures

**Comment by Fusezion** [comment]
> > I believe there was talk about whether they should be made as a structure but from what I remember we mostly put it down not fully sure what the reasons were
> 
> I see the reasoning here tbh, however arent structures able to be used as sections? i could be missing the ballpark completely but supporting both seems like something that structure is in range for, and if not that would be a nice imrpovment to structures

I'm aware that you can add the entry stuff as it's public but structures are top level events so they should be limited to the same area events, functions and commands are right?

**Comment by Fusezion** [comment]
Noo! we want theses

File: src/main/java/ch/njol/skript/sections/recipe/SecCookingRecipe.java [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+import ch.njol.skript.aliases.ItemType;
+import ch.njol.skript.config.EntryNode;
+import ch.njol.skript.config.Node;
+import ch.njol.skript.config.SectionNode;
+import ch.njol.skript.config.validate.SectionValidator;
+import ch.njol.skript.doc.Description;
+import ch.njol.skript.doc.Examples;
+import ch.njol.skript.doc.Name;
+import ch.njol.skript.doc.Since;
+import ch.njol.skript.lang.*;
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/sections/recipe/SecCookingRecipe.java:31]: wildcard imports

File: src/main/java/ch/njol/skript/sections/recipe/SecCookingRecipe.java [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+
+	static {
+			Skript.registerSection(SecCookingRecipe.class, "(create|add|register) [a] (0:blast furnace|1:campfire|2:furnace|3:smoker) recipe for %itemtype% with [the] key %string%");
+	}
+
+	private static SectionValidator validator = new SectionValidator()
+		.addEntry("ingredient", false)
+		.addEntry("cook time", false)
+		.addEntry("group", true)
+		.addEntry("xp", false);
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/sections/recipe/SecCookingRecipe.java:82]:
```suggestion
			.addEntry("ingredient", false)
			.addEntry("cook time", false)
			.addEntry("group", true)
			.addEntry("xp", false);
```

File: src/main/java/ch/njol/skript/sections/recipe/SecCookingRecipe.java [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+	private Expression<String> key;
+	private CookingRecipeType type;
+	private Expression<? extends ItemType> ingredient;
+	private Expression<? extends Timespan> cookTime;
+	private Expression<? extends Number> xp;
+	private Expression<ItemType> result;
+
+	@SuppressWarnings("unchecked")
+	@Override
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, SkriptParser.ParseResult parseResult, SectionNode sectionNode, List<TriggerItem> triggerItems) {
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/sections/recipe/SecCookingRecipe.java:98]:
```suggestion
	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult, SectionNode sectionNode, List<TriggerItem> triggerItems) {
```
import update required

File: src/main/java/ch/njol/skript/sections/recipe/SecCookingRecipe.java [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+
+	private Expression<String> key;
+	private CookingRecipeType type;
+	private Expression<? extends ItemType> ingredient;
+	private Expression<? extends Timespan> cookTime;
+	private Expression<? extends Number> xp;
+	private Expression<ItemType> result;
+
+	@SuppressWarnings("unchecked")
+	@Override
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/sections/recipe/SecCookingRecipe.java:97]:
```suggestion
	@Override
	@SuppressWarnings("unchecked")
```

File: src/main/java/ch/njol/skript/sections/recipe/SecCookingRecipe.java [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+
+	@Override
+	public String toString(Event event, boolean debug) {
+		return "create crafting recipe for " + result.toString(event, debug) + " with key " + key.toString(event, debug);
+	}
+
+	@Override
+	protected TriggerItem walk(Event event) {
+		execute(event);
+		return walk(event, false);
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/sections/recipe/SecCookingRecipe.java:122]: use super.walk

File: src/main/java/ch/njol/skript/sections/recipe/SecCookingRecipe.java [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+				break;
+			default:
+				throw new IllegalStateException();
+		}
+
+		try {
+			Bukkit.getServer().addRecipe(recipe);
+		} catch (IllegalStateException ignored) {
+			// Bukkit throws a IllegalStateException if a duplicate recipe is registered
+		}
```
Comment (Fusezion) [inline src/main/java/ch/njol/skript/sections/recipe/SecCookingRecipe.java:173]: Could we not just remove the duplicated recipe? I find that better than ignoring the error together

File: src/main/java/ch/njol/skript/sections/recipe/SecCookingRecipe.java [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+	}
+
+	private <T> Expression<? extends T> parse(EntryNode node, Class<T> parseAsClass) {
+		Node originalNode = getParser().getNode();
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/sections/recipe/SecCookingRecipe.java:177]: Nodes can be an entry node. Ensure the user hasn't defined a colon or equals here.

File: src/main/java/ch/njol/skript/sections/recipe/SecShapedRecipe.java [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+	static {
+		Skript.registerSection(SecShapedRecipe.class, "(create|add|register) [a] crafting recipe for %itemtype% with [the] key %string%");
+	}
+
+	private Expression<String> key;
+	private Expression<ItemType> ingredients;
+	private Expression<ItemType> result;
+
+	@SuppressWarnings("unchecked")
+	@Override
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/sections/recipe/SecShapedRecipe.java:70]:
```suggestion
	@Override
	@SuppressWarnings("unchecked")
```

File: src/main/java/ch/njol/skript/sections/recipe/SecShapedRecipe.java [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, SkriptParser.ParseResult parseResult, SectionNode sectionNode, List<TriggerItem> triggerItems) {
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/sections/recipe/SecShapedRecipe.java:71]:
```suggestion
	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult, SectionNode sectionNode, List<TriggerItem> triggerItems) {
```
import changes required

File: src/main/java/ch/njol/skript/sections/recipe/SecShapedRecipe.java [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+
+	@Override
+	public String toString(Event event, boolean debug) {
+		return "create crafting recipe for " + result.toString(event, debug) + " with key " + key.toString(event, debug);
+	}
+
+	@Override
+	protected TriggerItem walk(Event event) {
+		execute(event);
+		return walk(event, false);
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/sections/recipe/SecShapedRecipe.java:94]: super.walk for readability

File: src/main/java/ch/njol/skript/sections/recipe/SecCookingRecipe.java [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Stream;
+
+@Name("Cooking Recipe")
+@Description("Creates a cooking recipe")
+@Examples({
+	"create a crafting recipe for diamond named \"Dirty Diamond\" with the key \"dirty-diamond\":",
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/sections/recipe/SecCookingRecipe.java:65]: example doesn't reflect this class

**Review by TheLimeGlass — CHANGES_REQUESTED** [review]
There should also be a Structure for recipe registration. This will allow for parsing time registration. Makes it easier and simpler for users while also having the sections.

Tests can be added.

File: src/main/java/ch/njol/skript/sections/recipe/SecCookingRecipe.java (hunk 8 above)
Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/sections/recipe/SecCookingRecipe.java:173]: Configuration option to override is a possibility.

**Comment by sovdeeth** [comment]
State of the syntax at time of writing:
```
create a crafting recipe for diamond named "Dirty Diamond" with the key "dirty-diamond":
    shape:
        dirt, dirt and dirt
        dirt, diamond and dirt
        dirt, dirt and dirt
```

Personal thoughts on the existing syntax:
* I like that's it's a section. Most use cases of custom craft involve specific items that are rather unwieldy as literals (or impossible to write as literals), so the ability to use expressions and variables in the ingredients is, in my eyes, a key and essential feature.
* I like that it's easy to see the recipe at a glance. Laying it out in a 3x3 makes it much easier to read and see what the shape looks like, versus a 1x9 list.
* I'm not a huge fan of `shape:`. Since it's the only field in the section, I think it's rather superfluous. I get the idea that it's kind of a lead-in to writing the shape, and kinda takes the place of `with the shape "XYZ"`, but since it's the only thing you're going to put there anyway it feels more like just an unnecessary word to type than a helpful label. If there was a `result` field too, it'd make more sense in my opinion.
* I'm ok with the lack of an effect to start the shape lines. It's a new thing for skript, but I don't think it's bad. The section kind of acts like the effect for all the shape lines, making this more like a multi-line effect. I think it's intuitive and makes sense. However, I do have some alternatives if others find this too much of an upheaval.

Personal thoughts on additions:
* I agree that a structure version could be useful. However, if it doesn't allow expressions in its ingredients, as I understood is the case in a discord discussion, it'd be rather limited to just very simple recipes of generic material to generic material. This doesn't mean it's not worth adding, it's just that we should make sure it's well documented to ensure that the section's abilities and the structure's abilities are not confused.
* For the shape lines, my alternative suggestion:
```
crafting recipe:
    set [ingredients of] row 1 to x, y, z
    set [ingredients of] row 2 to x, y, z
    set [ingredients of] row 3 to x, y, z
```
This is mainly just to change the section as little as possible, but allow it to maintain the property of having an effect/section on each line of code. It could be extended with:
```
crafting recipe:
    set ingredients to x, y, z, a, b, c, 1, 2, 3
```
if desired.

### To summarize:
I like the current implementation of the syntax, minus the `shape:` bit, which I find a bit unnecessary. I think the lack of effect on the shape lines is perfectly fine, though I have an alternative if needed. In my experience, specific custom items are very commonly used in custom recipes, which means that if a structure was made, it'd be of somewhat limited use and would have to be well documented as to the differences with the section version. I don't think that's a reason not for it to exist, though.
All in all, I think this is a valuable addition to skript, and I like the current direction.

**Comment by TheLimeGlass** [comment]
Potentially add support for KnowledgeBookMeta? https://github.com/SkriptLang/Skript/issues/617

**Comment by TheLimeGlass** [comment]
Issue related https://github.com/SkriptLang/Skript/issues/5261

File: src/main/java/ch/njol/skript/sections/recipe/SecCookingRecipe.java (hunk 8 above)
Comment (Pikachu920) [reply src/main/java/ch/njol/skript/sections/recipe/SecCookingRecipe.java:175]: > Could we not just remove the duplicated recipe? I find that better than ignoring the error together

I'd have to write something that manually compares recipes to find duplicates. I would rather just rely on Bukkit's logic for detecting dupes so that bukkit is "in charge" of determining what is and isn't a duplicate recipe.

File: src/main/java/ch/njol/skript/sections/recipe/SecCookingRecipe.java [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+		}
+
+		if (this.group != null) {
+			String group = this.group.getSingle(event);
+			if (group != null)
+				recipe.setGroup(group);
+		}
+
+		try {
+			Bukkit.getServer().addRecipe(recipe);
```
Comment (Pikachu920) [inline src/main/java/ch/njol/skript/sections/recipe/SecCookingRecipe.java:173]: check this? it seems to return a boolean not throw an exception now that i look again

**Comment by sovdeeth** [comment]
Closing due to inactivity.
